### ISSUE TITLE
Media files get misclassified

### ISSUE THREAD

**Issue by meisanerd** [gerbera/gerbera#386]
I am running Gerbera from portage on Gentoo, version 1.2.0-r1, using MariaDB 10.1 (although the same issue arose with sqlite).  I switched to MariaDB so I could do external operations on the media library such as automatic playlist generation and other syncing operations for devices that don't support uPnP.
I have customized the import.js and playlists.js files.

diff /usr/share/gerbera/js/import.js import.js 
```
135,137d134
<     var chain = ['Video', 'All Video'];
<     addCdsObject(obj, createContainerChain(chain));
< 
141c138,140
<         chain = ['Video', 'Directories'];
---
>         dir.shift();
>         dir.shift();
>         chain = ['Video'];
```

diff /usr/share/gerbera/js/playlists.js playlists.js
```
74,75c74,75
< playlistChain = createContainerChain(['Playlists', 'All Playlists', playlist_title]);
< last_path = getLastPath(playlist.location);
---
> playlistChain = createContainerChain(['Playlists', playlist_title]);
> last_path = false;
```

My directory structure for music is music/[category]/[artist]/[album]/[songfile] where category is something like Christmas/Normal/Audiobook...

My directory structure for videos is videos/[category]/[first letter of show]/[show]/[season]/[videofile] where category is Movie/TV

When I run an import, it puts everything into the proper directory structure that I am looking for, while the import is running.  However, when the import finishes, some (not all) of the files get randomly put in the root, or disappear completely.

For audio, the all of the music shows up correctly categorized under Albums/Artists/etc, but the playlists seem to be a bit messed up.  It will randomly exclude some of the media files from the playlist (the playlist files are all valid .pls files, I sync these to different locations such as my phone and computers for offsite usage as well).

For video, it randomly takes some of the files out of the [category]/[letter] structure, and dumps them in the root under "Video".  Looking in the DB, it shows the parent category to be the "Video", not the letter it is supposed to be under.

I am not sure if there is logging that can be enabled to see what is going on, or if there is something wonky with the removal of the "All Playlists" and "All Video" categories (I removed them as going Playlists->All Playlists->Playlist I wanted was an extra unnecessary step, and I never used All Video as it takes forever to load on some of my devices and is hard to search).  The dir.shift are to remove the first 2 directories from the listing generation so it becomes Video->Category->Letter, rather than Video->Media->Video->Category in the listing.

**Comment by meisanerd** [comment]
As an update to this, I am now running 1.3.1 and the issue is still there.  I have tried it with my custom JS files, with the stock JS files, and with the virtual-layout set to builtin, and it still either hides files after a while, or misclassifies them.  This does not happen at initial import, but at a random period after the import has finished.  I was watching a movie the other night just after rebuilding and re-importing the database, and half way through the movie my Roku threw an error that the media file was not found and stopped playing it.  Going into Gerbera, the file had disappeared from the library, despite me changing absolutely nothing on the server.

This issue is starting to get really annoying, and it would be nice if it could be fixed, or at least if I had an idea of what was causing it.  Is there some sort of logging/debugging I can enable so that a potential issue would show up in the log file so I can track down what could be happening?

**Comment by meisanerd** [comment]
I added some print lines to the addVideo function of the import.js script, and noted that the first time through, it uses the full path to the set scan directory for the object_script_path, and subsequent times through, it uses the full path to the media item.  If I manually hard-code the object_script_path to be the scan directory, none of the files get mis-classified.  I am not really sure at this point where to dig into the server code to see what might be causing that.

**Comment by v00d00** [comment]
Thanks for the update, the last point of the differing paths after the first run is really interesting and a good starting point.

Can you give me some minimal instructions to replicate?

**Comment by meisanerd** [comment]
OK, so I am running Gentoo, last time I did a full emerge update world was about 1.5 weeks ago, I am running Gerbera 1.3.2 from ports.
> [ebuild   R   ~] net-misc/gerbera-1.3.2::gentoo  USE="curl exif ffmpeg javascript magic mysql taglib -debug -exiv2 -ffmpegthumbnailer -lastfm -libav -protocol-extensions -systemd"

Media files are stored in (and autoscan directory, every hour)
> /var/www/localhost/htdocs/media/files

Custom JS files for import are in
> /var/www/localhost/htdocs/media/gerbera-js

I have attached the 3 js files, as well as my config.xml for simplicity sake.  Most of the changes were done in import.js in the addVideo function (audio playlists are still having issues, haven't had the time to experiment with that one yet to see if I can fix it).  Also of note, I am doing a dir.shift() on the video files so when it creates the tree in Gerbera, it isn't video->Directories->files->{TV,Movies}, but video->{TV,Movies}, and I removed the "All Video" from the chain to simplify finding the things I wanted when using my DLNA clients.
[gerbera.tar.gz](https://github.com/gerbera/gerbera/files/3418417/gerbera.tar.gz)
